### PR TITLE
Improve test coverage

### PR DESCRIPTION
**Pull request by mcarton**
This adds a script `util/cov.sh` to run [`kcov`](https://github.com/SimonKagstrom/kcov) on clippy.
I did not configure travis to use because it takes some time (~15min on my machine), and uses a lot of memory. Also some results are weird:
* `kcov` does not seem to like rust's `match` very much;
* `kcov` definitely does not like our `it_let_chain!` macro (it only considers the first line, and we have a lot of those);
* how is that even possible?:
screenshot from 2016-10-02 15-17-56

Anyway, it did show some stuffs that we were not testing but could easily do: reported coverage goes up from 90.9% to 91.4% :tada: :balloon: 

For fun,^w^w here are the reports [before](https://github.com/Manishearth/rust-clippy/files/504958/before.zip) and [after](https://github.com/Manishearth/rust-clippy/files/504957/after.zip).

### CODE REVIEW

**Comment by Manishearth** [comment]
This is amazing work! Will review tomorrow.

**Comment by Manishearth** [comment]
We could let kcov only be run on master, fwiw. We don't care about travis cycle times on master (do we?)

**Comment by mcarton** [comment]
> do we?

Don't think so.

**Comment by llogiq** [comment]
We could also use coveralls?

**Comment by mcarton** [comment]
If we integrate it in travis, yes. `kcov` supports that.

File: tests/compile-fail/conf_bad_toml.toml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+fn this_is_obviously(not: a, toml: file) {
```
Comment (Manishearth) [inline tests/compile-fail/conf_bad_toml.toml]: Could we place this in `tests/aux`?

File: tests/compile-fail/conf_bad_type.toml [hunk 2]
```diff
@@ -0,0 +1 @@
+blacklisted-names = 42
```
Comment (Manishearth) [inline tests/compile-fail/conf_bad_type.toml]: ditto regarding `aux`

File: util/cov.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+tmpdir=$(mktemp -d)
+
+cargo test --no-run --verbose
+
+for t in "${tests[@]}"; do
+  kcov \
+    --verify \
+    --include-path="$(pwd)/src","$(pwd)/clippy_lints/src" \
+    "$tmpdir/$t" \
+    cargo test --test "$t"
```
Comment (Manishearth) [inline util/cov.sh]: wonder if there's a better way of doing this. perhaps we could turn off the compiletest tests with an env var, and then just run `cargo test`?

**Comment by Manishearth** [comment]
r=me when ready

**Comment by Manishearth** [comment]
Actually, looks ready now 😄

**Pull request merged**